test/python/circuit/classical/test_expr_constructors.py

```diff
@@ -610,3 +621,311 @@ def test_shift_forbidden(self, function):
+                expr.Cast(
+                    expr.Value(12, types.Uint(4)),
+                    types.Uint(8),
+                    implicit=False,
+                ),
+                types.Uint(8),
+            ),
+        )
+        self.assertTrue(expr.mul(expr.lift(12, types.Uint(8)), expr.lift(12)).const)
+
+        self.assertEqual(
+            expr.mul(expr.lift(12.0, types.Float()), expr.lift(12.0)),
+            expr.Binary(
+                expr.Binary.Op.MUL,
+                expr.Value(12.0, types.Float()),
+                expr.Value(12.0, types.Float()),
+                types.Float(),
+            ),
+        )
+        self.assertTrue(expr.mul(expr.lift(12.0, types.Float()), expr.lift(12.0)).const)
+
+        self.assertEqual(
+            expr.mul(Duration.ms(1000), 2.0),
+            expr.Binary(
+                expr.Binary.Op.MUL,
+                expr.Value(Duration.ms(1000), types.Duration()),
+                expr.Value(2.0, types.Float()),
+                types.Duration(),
+            ),
+        )
+        self.assertTrue(expr.mul(Duration.ms(1000), 2.0).const)
+
+        self.assertEqual(
+            expr.mul(2.0, Duration.ms(1000)),
+            expr.Binary(
+                expr.Binary.Op.MUL,
+                expr.Value(2.0, types.Float()),
+                expr.Value(Duration.ms(1000), types.Duration()),
+                types.Duration(),
+            ),
+        )
+        self.assertTrue(expr.mul(2.0, Duration.ms(1000)).const)
+
+        self.assertEqual(
+            expr.mul(2, Duration.ms(1000)),
+            expr.Binary(
+                expr.Binary.Op.MUL,
+                expr.Value(2, types.Uint(2)),
+                expr.Value(Duration.ms(1000), types.Duration()),
+                types.Duration(),
+            ),
+        )
+        self.assertTrue(expr.mul(2, Duration.ms(1000)).const)
+
+    def test_mul_forbidden(self):
+        with self.assertRaisesRegex(TypeError, "invalid types"):
+            expr.mul(Clbit(), ClassicalRegister(3, "c"))
+        with self.assertRaisesRegex(TypeError, "invalid types"):
+            expr.mul(ClassicalRegister(3, "c"), False)
+        with self.assertRaisesRegex(TypeError, "invalid types"):
+            expr.mul(Clbit(), Clbit())
+        with self.assertRaisesRegex(TypeError, "invalid types"):
+            expr.mul(0xFFFF, 2.0)
+        with self.assertRaisesRegex(TypeError, "invalid types"):
+            expr.mul(255.0, 1)
+        with self.assertRaisesRegex(TypeError, "cannot multiply two durations"):
+            expr.mul(Duration.dt(1000), Duration.dt(1000))
+
+        # Multiply timing expressions by non-const floats:
+        non_const_float = expr.Var.new("a", types.Float())
+        with self.assertRaisesRegex(ValueError, "would result in a non-const"):
+            expr.mul(Duration.dt(1000), non_const_float)
+        with self.assertRaisesRegex(ValueError, "would result in a non-const"):
+            expr.mul(non_const_float, Duration.dt(1000))
```

Comment:
In a world where we have a use and a way of functioning with runtime-specified `delay`, I can imagine us relaxing this restriction, but I think it's good to assert right now.